File: android/src/main/java/com/bugfender/sdk/capacitor/BugfenderPlugin.java
package com.bugfender.sdk.capacitor;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import androidx.activity.result.ActivityResult;
import androidx.annotation.Nullable;
import com.bugfender.sdk.Bugfender;
import com.bugfender.sdk.LogLevel;
import com.bugfender.sdk.ui.FeedbackActivity;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.net.URL;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "Bugfender")
public class BugfenderPlugin extends Plugin {

  private static final String OBFUSCATE_REQUEST_EVENT =
    "BugfenderObfuscateNetworkRequest";
  private static final String OBFUSCATE_RESPONSE_EVENT =
    "BugfenderObfuscateNetworkResponse";

  private static class PendingObfuscation {

    final CountDownLatch latch = new CountDownLatch(1);
    final AtomicReference<JSObject> result = new AtomicReference<>();
  }

  private final ConcurrentHashMap<String, PendingObfuscation> pendingObfuscations = new ConcurrentHashMap<>();

  @PluginMethod
  public void init(PluginCall call) {
    // region before init
    String deviceName = call.getString("deviceName");
    if (deviceName != null) {
      Bugfender.overrideDeviceName(deviceName);
    }

    String apiURL = call.getString("apiURL");
    if (apiURL != null) {
      Bugfender.setApiUrl(apiURL);
    }

    String baseURL = call.getString("baseURL");
    if (baseURL != null) {
      Bugfender.setBaseUrl(baseURL);
    }
    // endregion before init

    boolean debug = call.getBoolean(
      "debug",
      call.getBoolean("printToConsole", false)
    );
    Bugfender.init(getContext(), call.getString("appKey"), debug);

    // region after init
    Integer maximumLocalStorageSize = call.getInt("maximumLocalStorageSize");
    Bugfender.setMaximumLocalStorageSize(maximumLocalStorageSize);

    boolean enableLogcatLogging = call.getBoolean("enableLogcatLogging", false);
    if (enableLogcatLogging) {
      Bugfender.enableLogcatLogging();
    }

    boolean logUIEvents = call.getBoolean("logUIEvents", false);
    if (logUIEvents) {
      Bugfender.enableUIEventLogging(getActivity().getApplication());
    }

    boolean registerErrorHandler = call.getBoolean(
      "registerErrorHandler",
      false
    );
    if (registerErrorHandler) {
      Bugfender.enableCrashReporting();
    }

    if (Boolean.TRUE.equals(call.getBoolean("networkLoggingEnabled", false))) {
      Bugfender.setNetworkLoggingEnabled(true);
    }
    if (
      Boolean.TRUE.equals(call.getBoolean("networkLoggingCaptureBodies", false))
    ) {
      Bugfender.setNetworkLoggingCaptureBodies(true);
    }
    if (
      Boolean.TRUE.equals(
        call.getBoolean("networkLoggingCaptureErrorResponseBodies", false)
      )
    ) {
      Bugfender.setNetworkLoggingCaptureErrorResponseBodies(true);
    }
    // endregion after init

    call.resolve();
  }

  @PluginMethod
  public void forceSendOnce(PluginCall call) {
    Bugfender.forceSendOnce();
    call.resolve();
  }

  @PluginMethod
  public void getDeviceURL(PluginCall call) {
    URL url = Bugfender.getDeviceUrl();
    if (url != null) {
      JSObject response = new JSObject();
      response.put("url", url.toString());
      call.resolve(response);
    } else {
      call.reject(
        "Bugfender SDK is not initialized. You should call first to the method Bugfender.init()"
      );
    }
  }

  @PluginMethod
  public void getSessionURL(PluginCall call) {
    URL url = Bugfender.getSessionUrl();
    if (url != null) {
      JSObject response = new JSObject();
      response.put("url", url.toString());
      call.resolve(response);
    } else {
      call.reject(
        "Bugfender SDK is not initialized. You should call first to the method Bugfender.init()"
      );
    }
  }

  @PluginMethod
  public void getUserFeedback(PluginCall call) {
    String title = call.getString("title");
    String hint = call.getString("hint");
    String subjectHint = call.getString("subjectPlaceholder");
    String messageHint = call.getString("feedbackPlaceholder");
    String sendButtonText = call.getString("submitLabel");
    startActivityForResult(
      call,
      Bugfender.getUserFeedbackActivityIntent(
        getActivity().getApplication(),
        title,
        hint,
        subjectHint,
        messageHint,
        sendButtonText
      ),
      "getUserFeedbackResult"
    );
  }

  @ActivityCallback
  private void getUserFeedbackResult(PluginCall call, ActivityResult result) {
    if (call == null) {
      return;
    }
    if (result.getResultCode() == Activity.RESULT_OK) {
      JSObject response = new JSObject();
      response.put(
        "url",
        result.getData().getStringExtra(FeedbackActivity.RESULT_FEEDBACK_URL)
      );
      call.resolve(response);
    } else {
      call.reject("Feedback not sent");
    }
  }

  @PluginMethod
  public void log(PluginCall call) {
    Bugfender.d("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void warn(PluginCall call) {
    Bugfender.w("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void error(PluginCall call) {
    Bugfender.e("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void trace(PluginCall call) {
    Bugfender.t("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void info(PluginCall call) {
    Bugfender.i("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void fatal(PluginCall call) {
    Bugfender.f("", call.getString("text"));
    call.resolve();
  }

  @PluginMethod
  public void removeDeviceKey(PluginCall call) {
    Bugfender.removeDeviceKey(call.getString("key"));
    call.resolve();
  }

  @PluginMethod
  public void sendCrash(PluginCall call) {
    URL url = Bugfender.sendCrash(
      call.getString("title"),
      call.getString("text")
    );
    if (url != null) {
      JSObject response = new JSObject();
      response.put("url", url.toString());
      call.resolve(response);
    } else {
      call.reject(
        "Bugfender SDK is not initialized. You should call first to the method Bugfender.init()"
      );
    }
  }

  @PluginMethod
  public void sendIssue(PluginCall call) {
    URL url = Bugfender.sendIssue(
      call.getString("title"),
      call.getString("text")
    );
    if (url != null) {
      JSObject response = new JSObject();
      response.put("url", url.toString());
      call.resolve(response);
    } else {
      call.reject(
        "Bugfender SDK is not initialized. You should call first to the method Bugfender.init()"
      );
    }
  }

  @PluginMethod
  public void sendLog(PluginCall call) {
    Bugfender.log(
      call.getInt("line"),
      call.getString("method"),
      call.getString("file"),
      parseLogLevel(call.getInt("level")),
      call.getString("tag"),
      call.getString("text")
    );
  }

  private static LogLevel parseLogLevel(int logLevel) {
    switch (logLevel) {
      case 3:
        return LogLevel.Trace;
      case 4:
        return LogLevel.Info;
      case 5:
        return LogLevel.Fatal;
      case 1:
        return LogLevel.Warning;
      case 2:
        return LogLevel.Error;
      case 0:
      default:
        return LogLevel.Debug;
    }
  }

  @PluginMethod
  public void sendUserFeedback(PluginCall call) {
    URL url = Bugfender.sendUserFeedback(
      call.getString("title"),
      call.getString("text")
    );
    if (url != null) {
      JSObject response = new JSObject();
      response.put("url", url.toString());
      call.resolve(response);
    } else {
      call.reject(
        "Bugfender SDK is not initialized. You should call first to the method Bugfender.init()"
      );
    }
  }

  @PluginMethod
  public void setDeviceBoolean(PluginCall call) {
    String key = call.getString("key");
    Boolean value = call.getBoolean("value");

    Bugfender.setDeviceBoolean(key, value);
    call.resolve();
  }

  @PluginMethod
  public void setDeviceString(PluginCall call) {
    String key = call.getString("key");
    String value = call.getString("value");

    Bugfender.setDeviceString(key, value);
    call.resolve();
  }

  @PluginMethod
  public void setDeviceInteger(PluginCall call) {
    String key = call.getString("key");
    Integer value = call.getInt("value");

    Bugfender.setDeviceInteger(key, value);
    call.resolve();
  }

  @PluginMethod
  public void setDeviceFloat(PluginCall call) {
    String key = call.getString("key");
    Float value = call.getFloat("value");

    Bugfender.setDeviceFloat(key, value);
    call.resolve();
  }

  @PluginMethod
  public void setForceEnabled(PluginCall call) {
    Bugfender.setForceEnabled(call.getBoolean("state"));
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingEnabled(PluginCall call) {
    Bugfender.setNetworkLoggingEnabled(
      Boolean.TRUE.equals(call.getBoolean("enabled", false))
    );
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingCaptureBodies(PluginCall call) {
    Bugfender.setNetworkLoggingCaptureBodies(
      Boolean.TRUE.equals(call.getBoolean("capture", false))
    );
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingCaptureErrorResponseBodies(PluginCall call) {
    Bugfender.setNetworkLoggingCaptureErrorResponseBodies(
      Boolean.TRUE.equals(call.getBoolean("capture", false))
    );
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingURLFilter(PluginCall call) {
    Bugfender.setNetworkLoggingURLFilter(
      toStringList(call.getArray("allowlist")),
      toStringList(call.getArray("denylist"))
    );
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingMaxRequestsPerMinute(PluginCall call) {
    if (!call.getData().has("count") || call.getData().isNull("count")) {
      Bugfender.setNetworkLoggingMaxRequestsPerMinute(null);
    } else {
      Bugfender.setNetworkLoggingMaxRequestsPerMinute(call.getInt("count"));
    }
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingRequestObfuscationHandlerEnabled(
    PluginCall call
  ) {
    setObfuscationHandler(
      "setNetworkLoggingRequestObfuscationHandler",
      Boolean.TRUE.equals(call.getBoolean("enabled", false))
        ? createRequestObfuscationHandler()
        : null
    );
    call.resolve();
  }

  @PluginMethod
  public void setNetworkLoggingResponseObfuscationHandlerEnabled(
    PluginCall call
  ) {
    setObfuscationHandler(
      "setNetworkLoggingResponseObfuscationHandler",
      Boolean.TRUE.equals(call.getBoolean("enabled", false))
        ? createResponseObfuscationHandler()
        : null
    );
    call.resolve();
  }

  @PluginMethod
  public void completeNetworkObfuscation(PluginCall call) {
    String requestId = call.getString("requestId");
    if (requestId == null) {
      call.resolve();
      return;
    }
    PendingObfuscation pending = pendingObfuscations.get(requestId);
    if (pending != null) {
      pending.result.set(call.getObject("result"));
      pending.latch.countDown();
    }
    call.resolve();
  }

  /**
   * Android SDK 4.0.1 ships R8-obfuscated handler types. Resolve them via
   * reflection so this plugin compiles against the published Maven artifact.
   */
  private void setObfuscationHandler(String methodName, Object handler) {
    try {
      Method setter = findBugfenderMethod(methodName, 1);
      if (setter == null) {
        return;
      }
      setter.invoke(null, handler);
    } catch (Exception ignored) {
      // Optional API; ignore if unavailable.
    }
  }

  @Nullable
  private static Method findBugfenderMethod(String name, int paramCount) {
    for (Method method : Bugfender.class.getMethods()) {
      if (
        name.equals(method.getName()) &&
        method.getParameterTypes().length == paramCount
      ) {
        return method;
      }
    }
    return null;
  }

  @Nullable
  private Object createRequestObfuscationHandler() {
    Method setter = findBugfenderMethod(
      "setNetworkLoggingRequestObfuscationHandler",
      1
    );
    if (setter == null) {
      return null;
    }
    Class<?> handlerType = setter.getParameterTypes()[0];
    return Proxy.newProxyInstance(
      handlerType.getClassLoader(),
      new Class<?>[] { handlerType },
      (proxy, method, args) -> {
        if (method.getDeclaringClass() == Object.class) {
          return invokeObjectMethod(proxy, method, args);
        }
        if (args == null || args.length < 3) {
          return null;
        }
        String url = args[0] instanceof String ? (String) args[0] : "";
        @SuppressWarnings("unchecked")
        Map<String, String> headers = args[1] instanceof Map
          ? (Map<String, String>) args[1]
          : new HashMap<>();
        String body = args[2] instanceof String ? (String) args[2] : null;

        JSObject payload = new JSObject();
        payload.put("url", url != null ? url : "");
        payload.put("headers", toJSObject(headers));
        payload.put("body", body);

        JSObject response = invokeJsObfuscation(
          OBFUSCATE_REQUEST_EVENT,
          payload
        );
        String obfuscatedUrl = url;
        Map<String, String> obfuscatedHeaders = headers;
        String obfuscatedBody = body;
        if (response != null) {
          if (response.has("url") && !response.isNull("url")) {
            obfuscatedUrl = response.getString("url");
          }
          obfuscatedHeaders =
            headersFromJSObject(response.getJSObject("headers"));
          obfuscatedBody = null;
          if (response.has("body") && !response.isNull("body")) {
            obfuscatedBody = response.getString("body");
          }
        }
        return newNetworkData(
          method.getReturnType(),
          obfuscatedUrl,
          obfuscatedHeaders,
          obfuscatedBody
        );
      }
    );
  }

  @Nullable
  private Object createResponseObfuscationHandler() {
    Method setter = findBugfenderMethod(
      "setNetworkLoggingResponseObfuscationHandler",
      1
    );
    if (setter == null) {
      return null;
    }
    Class<?> handlerType = setter.getParameterTypes()[0];
    return Proxy.newProxyInstance(
      handlerType.getClassLoader(),
      new Class<?>[] { handlerType },
      (proxy, method, args) -> {
        if (method.getDeclaringClass() == Object.class) {
          return invokeObjectMethod(proxy, method, args);
        }
        if (args == null || args.length < 2) {
          return null;
        }
        @SuppressWarnings("unchecked")
        Map<String, String> headers = args[0] instanceof Map
          ? (Map<String, String>) args[0]
          : new HashMap<>();
        String body = args[1] instanceof String ? (String) args[1] : null;

        JSObject payload = new JSObject();
        payload.put("headers", toJSObject(headers));
        payload.put("body", body);

        JSObject response = invokeJsObfuscation(
          OBFUSCATE_RESPONSE_EVENT,
          payload
        );
        Map<String, String> obfuscatedHeaders = headers;
        String obfuscatedBody = body;
        if (response != null) {
          obfuscatedHeaders =
            headersFromJSObject(response.getJSObject("headers"));
          obfuscatedBody = null;
          if (response.has("body") && !response.isNull("body")) {
            obfuscatedBody = response.getString("body");
          }
        }
        return newNetworkData(
          method.getReturnType(),
          null,
          obfuscatedHeaders,
          obfuscatedBody
        );
      }
    );
  }

  @Nullable
  private static Object newNetworkData(
    Class<?> type,
    @Nullable String url,
    Map<String, String> headers,
    @Nullable String body
  ) throws Exception {
    for (Constructor<?> constructor : type.getConstructors()) {
      Class<?>[] params = constructor.getParameterTypes();
      if (
        params.length == 3 &&
        params[0] == String.class &&
        Map.class.isAssignableFrom(params[1]) &&
        params[2] == String.class
      ) {
        return constructor.newInstance(url, headers, body);
      }
      if (
        params.length == 2 &&
        Map.class.isAssignableFrom(params[0]) &&
        params[1] == String.class
      ) {
        return constructor.newInstance(headers, body);
      }
    }
    return null;
  }

  private static Object invokeObjectMethod(
    Object proxy,
    Method method,
    @Nullable Object[] args
  ) {
    String name = method.getName();
    if ("toString".equals(name)) {
      return "BugfenderNetworkObfuscationHandlerProxy";
    }
    if ("hashCode".equals(name)) {
      return System.identityHashCode(proxy);
    }
    if ("equals".equals(name)) {
      return proxy == (args != null && args.length > 0 ? args[0] : null);
    }
    return null;
  }

  @Nullable
  private JSObject invokeJsObfuscation(String eventName, JSObject body) {
    // Avoid deadlocking the UI thread while waiting for JS.
    if (Looper.myLooper() == Looper.getMainLooper()) {
      return null;
    }

    String requestId = UUID.randomUUID().toString();
    PendingObfuscation pending = new PendingObfuscation();
    pendingObfuscations.put(requestId, pending);
    body.put("requestId", requestId);

    new Handler(Looper.getMainLooper())
      .post(() -> notifyListeners(eventName, body));

    try {
      if (!pending.latch.await(3, TimeUnit.SECONDS)) {
        pendingObfuscations.remove(requestId);
        return null;
      }
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      pendingObfuscations.remove(requestId);
      return null;
    }

    pendingObfuscations.remove(requestId);
    return pending.result.get();
  }

  private static JSObject toJSObject(@Nullable Map<String, String> headers) {
    JSObject map = new JSObject();
    if (headers == null) {
      return map;
    }
    for (Map.Entry<String, String> entry : headers.entrySet()) {
      if (entry.getKey() != null) {
        map.put(
          entry.getKey(),
          entry.getValue() != null ? entry.getValue() : ""
        );
      }
    }
    return map;
  }

  private static Map<String, String> headersFromJSObject(
    @Nullable JSObject map
  ) {
    Map<String, String> result = new HashMap<>();
    if (map == null) {
      return result;
    }
    Iterator<String> keys = map.keys();
    while (keys.hasNext()) {
      String key = keys.next();
      try {
        if (map.isNull(key)) {
          result.put(key, "");
        } else {
          Object value = map.get(key);
          result.put(key, value != null ? String.valueOf(value) : "");
        }
      } catch (JSONException ignored) {
        result.put(key, "");
      }
    }
    return result;
  }

  @Nullable
  private static List<String> toStringList(@Nullable JSArray array) {
    if (array == null) {
      return null;
    }
    List<String> list = new ArrayList<>();
    try {
      for (int i = 0; i < array.length(); i++) {
        Object value = array.get(i);
        if (value != null && value != JSONObject.NULL) {
          list.add(String.valueOf(value));
        }
      }
    } catch (JSONException ignored) {
      // Return whatever we collected.
    }
    return list;
  }
}

File: ios/Plugin/BugfenderPlugin.swift
import Foundation
import Capacitor
import BugfenderSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(BugfenderPlugin)
// swiftlint:disable:next type_body_length
public class BugfenderPlugin: CAPPlugin {
    private var pendingObfuscations: [String: PendingObfuscation] = [:]
    private let pendingObfuscationsQueue = DispatchQueue(label: "com.bugfender.capacitor.obfuscation")

    @objc func `init`(_ call: CAPPluginCall) {
        // MARK: before init
        if let deviceName = call.getString("deviceName") {
            Bugfender.overrideDeviceName(deviceName)
        }

        if let apiURL = URL(string: call.getString("apiURL", "")) {
            Bugfender.setApiURL(apiURL)
        }

        if let baseURL = URL(string: call.getString("baseURL", "")) {
            Bugfender.setBaseURL(baseURL)
        }

        // MARK: init
        Bugfender.activateLogger(call.getString("appKey")!)

        // MARK: after init
        if let maximumLocalStorageSize = call.getInt("maximumLocalStorageSize") {
            Bugfender.setMaximumLocalStorageSize(UInt(maximumLocalStorageSize))
        }

        let logUIEvents = call.getBool("logUIEvents", false)
        if logUIEvents {
            Bugfender.enableUIEventLogging()
        }

        let registerErrorHandler = call.getBool("registerErrorHandler", false)
        if registerErrorHandler {
            Bugfender.enableCrashReporting()
        }

        if call.getBool("networkLoggingEnabled", false) {
            Bugfender.setNetworkLoggingEnabled(true)
        }
        if call.getBool("networkLoggingCaptureBodies", false) {
            Bugfender.setNetworkLoggingCaptureBodies(true)
        }
        if call.getBool("networkLoggingCaptureErrorResponseBodies", false) {
            Bugfender.setNetworkLoggingCaptureErrorResponseBodies(true)
        }

        call.resolve()
    }

    @objc func forceSendOnce(_ call: CAPPluginCall) {
        Bugfender.forceSendOnce()
        call.resolve()
    }

    @objc func getDeviceURL(_ call: CAPPluginCall) {
        let url = Bugfender.deviceIdentifierUrl()
        if let url = url?.absoluteString {
            call.resolve(["url": url])
        } else {
            call.reject("Bugfender SDK is not initialized. You should call first to the method Bugfender.init()")
        }
    }

    @objc func getSessionURL(_ call: CAPPluginCall) {
        let url = Bugfender.sessionIdentifierUrl()
        if let url = url?.absoluteString {
            call.resolve(["url": url])
        } else {
            call.reject("Bugfender SDK is not initialized. You should call first to the method Bugfender.init()")
        }
    }

    @objc func getUserFeedback(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let feedbackViewController = BFUserFeedbackNavigationController.userFeedbackViewController(
                withTitle: call.getString("title")!,
                hint: call.getString("hint")!,
                subjectPlaceholder: call.getString("subjectPlaceholder")!,
                messagePlaceholder: call.getString("feedbackPlaceholder")!,
                sendButtonTitle: call.getString("submitLabel")!,
                cancelButtonTitle: call.getString("closeLabel")!) { (sent, feedbackUrl) in
                if sent == true,
                   let url = feedbackUrl?.absoluteString {
                    call.resolve(["url": url])
                } else {
                    call.reject("Feedback not sent")
                }
            }

            self.bridge?.viewController?.present(feedbackViewController, animated: true, completion: nil)
        }
    }

    @objc func log(_ call: CAPPluginCall) {
        BFLog(call.getString("text")!)
        call.resolve()
    }

    @objc func warn(_ call: CAPPluginCall) {
        BFLogWarn(call.getString("text")!)
        call.resolve()
    }

    @objc func error(_ call: CAPPluginCall) {
        BFLogErr(call.getString("text")!)
        call.resolve()
    }

    @objc func trace(_ call: CAPPluginCall) {
        BFLogTrace(call.getString("text")!)
        call.resolve()
    }

    @objc func info(_ call: CAPPluginCall) {
        BFLogInfo(call.getString("text")!)
        call.resolve()
    }

    @objc func fatal(_ call: CAPPluginCall) {
        BFLogFatal(call.getString("text")!)
        call.resolve()
    }

    @objc func removeDeviceKey(_ call: CAPPluginCall) {
        Bugfender.removeDeviceKey(call.getString("key")!)
        call.resolve()
    }

    @objc func sendCrash(_ call: CAPPluginCall) {
        let url = Bugfender.sendCrash(withTitle: call.getString("title")!, text: call.getString("text")!)
        if let url = url?.absoluteString {
            call.resolve(["url": url])
        } else {
            call.reject("Bugfender SDK is not initialized. You should call first to the method Bugfender.init()")
        }
    }

    @objc func sendIssue(_ call: CAPPluginCall) {
        let url = Bugfender.sendIssueReturningUrl(withTitle: call.getString("title")!, text: call.getString("text")!)
        if let url = url?.absoluteString {
            call.resolve(["url": url])
        } else {
            call.reject("Bugfender SDK is not initialized. You should call first to the method Bugfender.init()")
        }
    }

    @objc func sendLog(_ call: CAPPluginCall) {
        Bugfender.log(
            lineNumber: call.getInt("line")!,
            method: call.getString("method")!,
            file: call.getString("file")!,
            level: parseLogLevel(call.getInt("level")!),
            tag: call.getString("tag"),
            message: call.getString("text")!
        )
        call.resolve()
    }

    private func parseLogLevel(_ logLevel: Int) -> BFLogLevel {
        switch logLevel {
        case 3:
            return BFLogLevel.trace
        case 4:
            return BFLogLevel.info
        case 5:
            return BFLogLevel.fatal
        case 1:
            return BFLogLevel.warning
        case 2:
            return BFLogLevel.error
        default:
            return BFLogLevel.default
        }
    }

    @objc func sendUserFeedback(_ call: CAPPluginCall) {
        let url = Bugfender.sendUserFeedbackReturningUrl(withSubject: call.getString("title")!, message: call.getString("text")!)
        if let url = url?.absoluteString {
            call.resolve(["url": url])
        } else {
            call.reject("Bugfender SDK is not initialized. You should call first to the method Bugfender.init()")
        }
    }

    @objc func setDeviceBoolean(_ call: CAPPluginCall) {
        let key =  call.getString("key")!
        let value = call.getBool("value")!

        Bugfender.setDeviceBOOL(value, forKey: key)
        call.resolve()
    }

    @objc func setDeviceString(_ call: CAPPluginCall) {
        let key =  call.getString("key")!
        let value = call.getString("value")!

        Bugfender.setDeviceString(value, forKey: key)
        call.resolve()
    }

    @objc func setDeviceInteger(_ call: CAPPluginCall) {
        let key =  call.getString("key")!
        let value = call.getInt("value")!

        Bugfender.setDeviceInteger(UInt64(value), forKey: key)
        call.resolve()
    }

    @objc func setDeviceFloat(_ call: CAPPluginCall) {
        let key =  call.getString("key")!
        let value = call.getDouble("value")!

        Bugfender.setDeviceDouble(value, forKey: key)
        call.resolve()
    }

    @objc func setForceEnabled(_ call: CAPPluginCall) {
        Bugfender.setForceEnabled(call.getBool("state")!)
        call.resolve()
    }

    @objc func setNetworkLoggingEnabled(_ call: CAPPluginCall) {
        Bugfender.setNetworkLoggingEnabled(call.getBool("enabled", false))
        call.resolve()
    }

    @objc func setNetworkLoggingCaptureBodies(_ call: CAPPluginCall) {
        Bugfender.setNetworkLoggingCaptureBodies(call.getBool("capture", false))
        call.resolve()
    }

    @objc func setNetworkLoggingCaptureErrorResponseBodies(_ call: CAPPluginCall) {
        Bugfender.setNetworkLoggingCaptureErrorResponseBodies(call.getBool("capture", false))
        call.resolve()
    }

    @objc func setNetworkLoggingURLFilter(_ call: CAPPluginCall) {
        let allowlist = call.getArray("allowlist", String.self)
        let denylist = call.getArray("denylist", String.self)
        Bugfender.setNetworkLoggingURLFilter(allowlist: allowlist, denylist: denylist)
        call.resolve()
    }

    @objc func setNetworkLoggingMaxRequestsPerMinute(_ call: CAPPluginCall) {
        if call.getValue("count") == nil || call.getValue("count") is NSNull {
            Bugfender.setNetworkLoggingMaxRequestsPerMinute(nil)
        } else if let count = call.getInt("count") {
            Bugfender.setNetworkLoggingMaxRequestsPerMinute(count)
        } else {
            Bugfender.setNetworkLoggingMaxRequestsPerMinute(nil)
        }
        call.resolve()
    }

    @objc func setNetworkLoggingRequestObfuscationHandlerEnabled(_ call: CAPPluginCall) {
        if call.getBool("enabled", false) {
            installRequestObfuscationHandler()
        } else {
            Bugfender.setNetworkLoggingRequestObfuscationHandler(nil)
        }
        call.resolve()
    }

    @objc func setNetworkLoggingResponseObfuscationHandlerEnabled(_ call: CAPPluginCall) {
        if call.getBool("enabled", false) {
            installResponseObfuscationHandler()
        } else {
            Bugfender.setNetworkLoggingResponseObfuscationHandler(nil)
        }
        call.resolve()
    }

    @objc func completeNetworkObfuscation(_ call: CAPPluginCall) {
        guard let requestId = call.getString("requestId"), !requestId.isEmpty else {
            call.resolve()
            return
        }

        var pending: PendingObfuscation?
        pendingObfuscationsQueue.sync {
            pending = pendingObfuscations[requestId]
        }
        if let pending = pending {
            pending.response = call.getObject("result")
            pending.semaphore.signal()
        }
        call.resolve()
    }

    /// Example / verification helper: URLSession request so traffic appears as `bf_network`.
    @objc func sendInstrumentedNetworkRequest(_ call: CAPPluginCall) {
        let urlString = call.getString("url") ?? "https://example.com/"
        let httpMethod = (call.getString("method") ?? "GET").uppercased()
        let body = call.getString("body")
        let extraHeaders = call.getObject("headers") ?? [:]

        guard let url = URL(string: urlString) else {
            call.reject("Invalid URL")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = httpMethod
        request.timeoutInterval = 15.0

        var hasAuthorization = false
        for (key, value) in extraHeaders {
            if let stringValue = value as? String {
                request.setValue(stringValue, forHTTPHeaderField: key)
                if key.compare("Authorization", options: .caseInsensitive) == .orderedSame {
                    hasAuthorization = true
                }
            }
        }
        if !hasAuthorization {
            request.setValue("secret-token", forHTTPHeaderField: "Authorization")
        }

        if let body = body,
           !body.isEmpty,
           httpMethod == "POST" || httpMethod == "PUT" || httpMethod == "PATCH" {
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.httpBody = body.data(using: .utf8)
        }

        let task = URLSession.shared.dataTask(with: request) { _, response, error in
            if let error = error {
                call.reject(error.localizedDescription)
                return
            }
            let status: Int
            if let httpResponse = response as? HTTPURLResponse {
                status = httpResponse.statusCode
            } else {
                status = 0
            }
            DispatchQueue.main.async {
                Bugfender.forceSendOnce()
                call.resolve([
                    "status": status,
                    "shouldCapture": true,
                    "requestId": NSNull()
                ])
            }
        }
        task.resume()
    }

    private func installRequestObfuscationHandler() {
        Bugfender.setNetworkLoggingRequestObfuscationHandler { [weak self] url, headers, body in
            guard let self = self else {
                return (url: url, headers: headers, body: body)
            }

            let response = self.invokeJSObfuscation(
                eventName: "BugfenderObfuscateNetworkRequest",
                data: [
                    "url": url,
                    "headers": headers,
                    "body": body as Any
                ]
            )
            guard let response = response else {
                return (url: url, headers: headers, body: body)
            }

            let obfuscatedUrl = response["url"] as? String ?? url
            let obfuscatedHeaders = self.stringMap(from: response["headers"])
            let obfuscatedBody = response["body"] as? String
            return (url: obfuscatedUrl, headers: obfuscatedHeaders, body: obfuscatedBody)
        }
    }

    private func installResponseObfuscationHandler() {
        Bugfender.setNetworkLoggingResponseObfuscationHandler { [weak self] headers, body in
            guard let self = self else {
                return (headers: headers, body: body)
            }

            let response = self.invokeJSObfuscation(
                eventName: "BugfenderObfuscateNetworkResponse",
                data: [
                    "headers": headers,
                    "body": body as Any
                ]
            )
            guard let response = response else {
                return (headers: headers, body: body)
            }

            let obfuscatedHeaders = self.stringMap(from: response["headers"])
            let obfuscatedBody = response["body"] as? String
            return (headers: obfuscatedHeaders, body: obfuscatedBody)
        }
    }

    private func invokeJSObfuscation(eventName: String, data: [String: Any]) -> [String: Any]? {
        // Avoid deadlocking the platform/UI thread while waiting for JS.
        if Thread.isMainThread {
            return nil
        }

        let requestId = UUID().uuidString
        let pending = PendingObfuscation()
        pendingObfuscationsQueue.sync {
            pendingObfuscations[requestId] = pending
        }

        var payload = data
        payload["requestId"] = requestId

        DispatchQueue.main.async { [weak self] in
            self?.notifyListeners(eventName, data: payload)
        }

        let waitResult = pending.semaphore.wait(timeout: .now() + 3.0)
        pendingObfuscationsQueue.sync {
            pendingObfuscations.removeValue(forKey: requestId)
        }

        if waitResult == .timedOut {
            return nil
        }
        return pending.response
    }

    private func stringMap(from value: Any?) -> [String: String] {
        guard let raw = value as? [AnyHashable: Any] else {
            return [:]
        }
        var mapped: [String: String] = [:]
        for (key, entry) in raw {
            if entry is NSNull {
                mapped[String(describing: key)] = ""
            } else {
                mapped[String(describing: key)] = String(describing: entry)
            }
        }
        return mapped
    }
}

private final class PendingObfuscation {
    let semaphore = DispatchSemaphore(value: 0)
    var response: [String: Any]?
}
